File: image_processing/01_images/Python/01_imagePGM_class/01_image_class.py
"""01_image_class file.

File containing :class::ImagePGM
Class to represent a PGM image.


.. note:: LEnsE - Institut d'Optique - version 1.0

.. moduleauthor:: Julien VILLEMEJANE <julien.villemejane@institutoptique.fr>

"""

class ImagePGM:
    """
    Class to represent a PGM image.

    :param width: Width of the image.
    :type width: int
    :param height: Height of the image.
    :type height: int
    :param channels: Number of color values for each pixel of the image.
    :type channels: int
    :param max_gray_value: Value of the white color.
    :type max_gray_value: int
    :param magic_number: Type of the file.
    :type magic_number: str
    :param pixels: Value of each pixel.
    :type pixels: list
    """
    
    def __init__(self, filename):
        """
        Initialize the ImagePGM object.

        Parameters:
            filename (str): Name of the PGM file.
        """
        self.width = 0          # Width of the image
        self.height = 0         # Height of the image
        self.channels = 0       # Number of color values for each pixel of the image
        self.max_gray_value = 0 # Value of the white color
        self.magic_number = ""  # Type of the file
        self.pixels = []        # Value of each pixel
        self.readImagePGM(filename)
    
    def readImagePGM(self, filename):
        """
        Read a PGM image from a file and store its information.

        :param filename: Name of the PGM file.
        :type filename: str
        
        :return: True if the file was successfully read, False otherwise.
        :rtype: bool

        """
        try:
            with open(filename, 'r') as file:
                lines = file.readlines()
                self.magic_number = lines[0].strip()
                self.width, self.height = map(int, lines[1].split())
                self.max_gray_value = int(lines[2])
                self.pixels = [[int(value) for value in line.split()] for line in lines[3:]]
        except FileNotFoundError:
            print(f"Error: Unable to open file {filename}")
            return False
        return True

    def __str__(self):
        """
        Return a string representation of the ImagePGM object.

        :return: A string with PGM image informations
        :rtype: str
        """
        return f"Image PGM / Type: {self.magic_number}, (W,H) = ({self.width}, {self.height})"


# Main function
if __name__ == "__main__":
    # Read the input PGM image
    image = ImagePGM("../../_data/robot.pgm")

    print("Infos")
    print(image)

File: image_processing/01_images/Python/01_imagePGM_class/01_image_class.cpp
#include <iostream>
#include <string>
#include <vector>
#include <fstream> // Include <fstream> for file operations

class ImagePGM {
	public:
		/**
		 * @brief Construct a new ImagePGM object from a filename.
		 * @param filename Name of the file name of the image (constant).
		 */
		ImagePGM(const std::string& filename);
		
		/**
		 * @brief Open an image file.
		 * @param filename Name of the file of the image (constant).
		 * @return true if the file exists.
		 */
		bool readImagePGM(const std::string& filename);
		
		// Declaration of operator<< as a friend function
		friend std::ostream& operator<<(std::ostream& os, const ImagePGM& img);

	private:
		int width;             /**< Width of the image. */
		int height;            /**< Height of the image. */
		int channels;          /**< Number of color value for each pixel of the image. */
		int max_gray_value;    /**< Value of the white color. */
		std::string magic_number;  /**< Type of the file. */

		std::vector< std::vector<int> > pixels; /**< Value of each pixels. */

};

// Function to read a PGM image from a file
bool ImagePGM::readImagePGM(const std::string& filename) {
    std::ifstream file(filename.c_str());

    if (file.is_open()) {
        file >> this->magic_number >> this->width >> this->height >> this->max_gray_value;

        this->pixels.resize(this->height, std::vector<int>(this->width));

        for (int i = 0; i < this->height; ++i) {
            for (int j = 0; j < this->width; ++j) {
                file >> this->pixels[i][j];
            }
        }

        file.close();
    } else {
        std::cerr << "Error: Unable to open file " << filename << std::endl;
        return false;
    }

    return true;
}

// Constructor definition
ImagePGM::ImagePGM(const std::string& filename) {
    readImagePGM(filename);
}

// Overloading the << operator to print information about the object
std::ostream& operator<<(std::ostream& os, const ImagePGM& img) {
	os << "Image PGM / Type : " << img.magic_number << ", (W,H) = (" << img.width << "," << img.height << ')' << std::endl;
	return os;
}


int main(int argc, char** argv) {
    // Read the input PGM image
    ImagePGM image("../../_data/robot.pgm");
	
	std::cout << "Infos" << std::endl;
	std::cout << image << std::endl;

    return 0;
}
